### PR TITLE
Add VM File handling to qcore

### PR DESCRIPTION
**Pull request by jpa198**
Create simulation structure functions for Qp/Qs files and vm_params

### CODE REVIEW

File: qcore/vm_file.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        return method(vm_obj, *args)
+
+    return inner
+
+
+class VelocityModelFile:
+
+    _data = None
+
+    def __init__(self, nx, ny, nz, file_loc=None):
```
Comment (jasonmotha) [inline qcore/vm_file.py:25]: Default should have a read_only attribute

File: qcore/vm_file.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        if self.file_path is None:
+            raise ValueError(
+                "Must set file path. Or use new() to create a new file from scratch"
+            )
+
+        if exists(self.file_path):
+            self._data = np.fromfile(self.file_path, "<f4").reshape(
+                [self.ny, self.nz, self.nx]
+            )
+            self._data = np.swapaxes(self._data, 0, 2)
```
Comment (jasonmotha) [inline qcore/vm_file.py:61]: Either abstract the swapping axeses away or explain what's happening here.

File: qcore/vm_file.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        self._data = None
+
+    @_check_data
+    def get_value(self, x, y, z):
+        if self._data is None:
+            raise ValueError("Must open file first")
+        return self._data[x, y, z]
+
+    @_check_data
+    def set_value(self, value, x, y, z):
```
Comment (jasonmotha) [inline qcore/vm_file.py:92]: doc string. About changing data but not writing to file.

**Review by jasonmotha — CHANGES_REQUESTED** [review]

File: qcore/vm_file.py [hunk 4]
```diff
@@ -1,104 +1,243 @@
...
+def create_constant_vm_file(pert_f_location, npoints, value=1):
+    """
+    Creates a vm file with given size and uniform value
+    :param pert_f_location: The location to save the file to
+    :param npoints: The size of the file to make
+    :param value: The value to set the model to. Defaults to 1
+    """
+    # Doesn't matter what the dimensions are, we only need the total point count
+    with VelocityModelFile(npoints, 1, 1) as vmf:
+        vmf.set_values(np.ones(vmf.shape)*value)
```
[the remark this answers is not in the document]
Comment (jpa198) [reply qcore/vm_file.py]: TIL

**Review by jasonmotha — APPROVED** [review]

**Review by jasonmotha — APPROVED** [review]

**Comment by jpa198** [comment]
Even without the Qp/Qs stuff, I feel this is ready to go into the repo

**Comment by jasonmotha** [comment]
Agreed.

**Review by shot0829 — APPROVED** [review]

**Pull request merged**